### ISSUE TITLE
Weird transaction behaviour

### ISSUE THREAD

**Issue by karls** [nikita-volkov/hasql-transaction#8]
Hi Nikita,

I've been running hasql in production for a while now, very-very pleased with it! Thanks!

Today, while working on an unrelated feature, I discovered a weird anomaly in our data. Seemingly, occasionally a query in a transaction fails to commit its result, but the subsequent queries that rely on the result of the first one "doing the expected thing", work as if everything was fine.

Here's a code example (some domain-specific names changed):
```hs
let txn = do
      void $ Txn.query (childId, Processed) Child.setState
      children <- Txn.query parentId Parent.getChildren
      when (all (\c -> childState c == Processed) children) $ do
        Txn.query parentId Parent.markAsDone
        Txn.query parentId Parent.doSomeUnrelatedThing
    sess = Txn.run txn Txn.RepeatableRead Txn.Write
  in do Hasql.use pool sess
        someUnrelatedCleanupAction
```

The weird thing is, every once in a while I get a `Parent` that's marked as done, but one of the children is *not* in the `Processed` state, which according to the code block above should not happen. This is code in one of the workers and I can't see any errors or retries in the worker logs. I'm also waiting for query logs from our database vendor to see if there are any clues in there.

Do you see anything wrong with the above code? Could it be a bug? Am I misusing the library somehow?

**Comment by nikita-volkov** [comment]
Might be due to the `RepeatableRead` isolation level, because with it phantom reads can happen. Try setting it to `Serializable`.

P.S. Thanks for the feedback!

**Comment by karls** [comment]
The Postgres docs claim that phantom reads are not possible in Postgres -- https://www.postgresql.org/docs/9.6/static/transaction-iso.html#MVCC-ISOLEVEL-TABLE.

On the other hand, the last paragraph of the `RepeatableRead` section reads:
> For example, even a read only transaction at this level may see a control record updated to show that a batch has been completed but not see one of the detail records which is logically part of the batch because it read an earlier revision of the control record. Attempts to enforce business rules by transactions running at this isolation level are not likely to work correctly without careful use of explicit locks to block conflicting transactions.

Is it possible that I'm hitting this case? Either way, I'm going to try changing it to `Serializable` and see what happens.

**Comment by nikita-volkov** [comment]
I can't say for sure. Transactions are a very fragile matter from my experience.

If you want to resolve this for certain, I recommend reproducing the issue in an isolated setting. It would be best if it was implemented as a test-suite for "hasql-transaction".

**Comment by karls** [comment]
Yeah, agreed. It's just puzzling that somehow the `Child.setState` query seems to be rolled back, but also has a definite effect on the subsequent queries which get committed just fine.. 🤔

Anyway, yes, I think coming up with a test case for this is the best way to go. I'll see if I can spend a few hours today on this.

**Comment by karls** [comment]
Just a quick update. I started working on reproducing the behaviour in the hasql-transaction test suite. I've got a test set up, but it does not reflect accurately what happens in my application, so I'll need to tweak it.

I was going through my data again today and discovered something even more bizarre.
```
|  Parent ID  | Parent's "processed" timestamp|  Child ID    | Child's state | Child's "updated" timestamp   |
--------------------------------------------------------------------------------------------------------------
| <parent ID> │ 2017-04-05 13:49:30.005193+00 │ <child ID 1> │ Processed     │ 2017-04-05 13:45:54.710877+00 │
│ <parent ID> │ 2017-04-05 13:49:30.005193+00 │ <child ID 4> │ Processed     │ 2017-04-05 13:48:29.078849+00 │
│ <parent ID> │ 2017-04-05 13:49:30.005193+00 │ <child ID 2> │ Processed     │ 2017-04-05 13:46:53.101328+00 │
│ <parent ID> │ 2017-04-05 13:49:30.005193+00 │ <child ID 5> │ OnDisk        │ 2017-04-05 13:49:30.005193+00 │
│ <parent ID> │ 2017-04-05 13:49:30.005193+00 │ <child ID 4> │ Processed     │ 2017-04-05 13:47:32.134003+00 │
```

This is a simple join of the parent and its children. Whenever a child's state changes, it's "updated" column is also touched (with `now()`). And whenever all children are processed, the parent's "processed" column is filled in.
The suspicious row is the `OnDisk` one -- it should be `Processed` because the parent's "processed" column has a timestamp. However, notice how the parent's "processed" timestamp matches that of the last updated child's timestamp!

Anyway, I thought I'd share this, in case it starts ringing any bells for you.

**Comment by karls** [comment]
Ahhh. Perhaps unsurprisingly, the error was between the keyboard and the chair. 

I didn't get very far with reproducing the error, because I got the logs from our database vendor. As I cross-referenced the queries with the data and code I realised what the problem was. A query inside a web handler (far-far from that worker) that looked like was doing only `SELECT` queries was, under certain conditions, actually doing `UPDATE`-s as well. Doh!

In short, panic over. Thanks for your input! :-)
